llvm/lib/SYCLLowerIR/MutatePrintfAddrspace.cpp

```diff
@@ -0,0 +1,145 @@
+
+#include "llvm/SYCLLowerIR/MutatePrintfAddrspace.h"
+
+#include "llvm/Analysis/ValueTracking.h"
+#include "llvm/IR/IRBuilder.h"
+#include "llvm/IR/Instructions.h"
+#include "llvm/InitializePasses.h"
+
+using namespace llvm;
+
+namespace {
+constexpr unsigned ConstantAddrspaceID = 2;
+
+// Wrapper for the pass to make it working with the old pass manager
+class SYCLMutatePrintfAddrspaceLegacyPass : public ModulePass {
+public:
+  static char ID;
+  SYCLMutatePrintfAddrspaceLegacyPass() : ModulePass(ID) {
+    initializeSYCLMutatePrintfAddrspaceLegacyPassPass(
+        *PassRegistry::getPassRegistry());
+  }
+
+  // run the SYCLMutatePrintfAddrspace pass on the specified module
+  bool runOnModule(Module &M) override {
+    ModuleAnalysisManager MAM;
+    auto PA = Impl.run(M, MAM);
+    return !PA.areAllPreserved();
+  }
+
+private:
+  SYCLMutatePrintfAddrspacePass Impl;
+};
+
+Constant *getCASLiteral(IRBuilder<> &Builder, CallInst *CI) {
+  auto *Literal = cast<Constant>(CI->getArgOperand(0));
+  Type *LiteralType = Literal->getType();
+  // Generate/find a correct literal
+  auto *CASLiteralType = PointerType::get(LiteralType->getPointerElementType(),
+                                          ConstantAddrspaceID);
+  StringRef CASLiteralName(Literal->getName().str() + "._AS2");
+  return CI->getModule()->getOrInsertGlobal(
+      CASLiteralName, CASLiteralType, [&] {
+        StringRef LiteralValue;
+        getConstantStringInfo(Literal, LiteralValue);
+        GlobalVariable *GV = Builder.CreateGlobalString(
+            LiteralValue, CASLiteralName, ConstantAddrspaceID, CI->getModule());
+        GV->setLinkage(GlobalValue::LinkageTypes::InternalLinkage);
+        GV->setUnnamedAddr(GlobalValue::UnnamedAddr::None);
+        return GV;
+      });
+}
+
+FunctionCallee getCASPrintfFunction(Function &GenericASPrintfFunc,
+                                    Type *CASLiteralTy) {
+  auto *CASPrintfFuncTy = FunctionType::get(GenericASPrintfFunc.getReturnType(),
+                                            CASLiteralTy, /*isVarArg=*/true);
+  FunctionCallee CASPrintfFunc =
+      GenericASPrintfFunc.getParent()->getOrInsertFunction(
+          "_Z18__spirv_ocl_printfPU3AS2Kcz", CASPrintfFuncTy,
+          GenericASPrintfFunc.getAttributes());
+  auto *Callee = cast<Function>(CASPrintfFunc.getCallee());
+  Callee->setCallingConv(CallingConv::SPIR_FUNC);
+  Callee->setDSOLocal(true);
+  return CASPrintfFunc;
+}
+
+CallInst *buildCASPrintfCall(IRBuilder<> &Builder, FunctionCallee CASPrintfFunc,
+                             Constant *CASLiteral, CallInst *CI) {
+  SmallVector<Value *, 4> CallOperands(CI->data_operands_begin(),
+                                       CI->data_operands_end());
+  CallOperands[0] = CASLiteral;
+  Builder.SetInsertPoint(CI);
+  CallInst *CASCall = Builder.CreateCall(CASPrintfFunc, CallOperands,
+                                         CI->getName().str() + "._AS2");
```

Comment:
Directly using [`CallInst::Create`](https://llvm.org/doxygen/classllvm_1_1CallInst.html#acee7881ae4925e4eafef04d0939b7f2d) could simplify (or at least shorten) this code, I think.

Reply:
Done